src/sys/unix/selector/poll.rs

```diff
@@ -0,0 +1,518 @@
+use crate::{Interest, Token};
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::fmt::{Debug, Formatter};
+use std::os::unix::io::RawFd;
+use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+use std::sync::{Arc, Condvar, Mutex};
+use std::time::{Duration, Instant};
+use std::{fmt, io};
+
+/// Unique id for use as `SelectorId`.
+#[cfg(debug_assertions)]
+static NEXT_ID: AtomicUsize = AtomicUsize::new(1);
+
+#[cfg(target_os = "espidf")]
+type NotifyType = u64;
+
+#[cfg(not(target_os = "espidf"))]
+type NotifyType = u8;
+
+#[derive(Debug)]
+pub struct Selector {
+    state: Arc<SelectorState>,
+    /// Whether this selector currently has an associated waker.
+    #[cfg(debug_assertions)]
+    has_waker: AtomicBool,
+}
+
+impl Selector {
+    pub fn new() -> io::Result<Selector> {
+        let state = SelectorState::new()?;
+
+        Ok(Selector {
+            state: Arc::new(state),
+            #[cfg(debug_assertions)]
+            has_waker: AtomicBool::new(false),
+        })
+    }
+
+    pub fn try_clone(&self) -> io::Result<Selector> {
+        let state = self.state.clone();
+
+        Ok(Selector {
+            state,
+            #[cfg(debug_assertions)]
+            has_waker: AtomicBool::new(false),
+        })
+    }
+
+    pub fn select(&self, events: &mut Events, timeout: Option<Duration>) -> io::Result<()> {
+        self.state.select(events, timeout)
+    }
+
+    pub fn register(&self, fd: RawFd, token: Token, interests: Interest) -> io::Result<()> {
+        self.state.register(fd, token, interests)
+    }
+
+    pub fn reregister(&self, fd: RawFd, token: Token, interests: Interest) -> io::Result<()> {
+        self.state.reregister(fd, token, interests)
+    }
+
+    pub fn deregister(&self, fd: RawFd) -> io::Result<()> {
+        self.state.deregister(fd)
+    }
+
+    #[cfg(debug_assertions)]
+    pub fn register_waker(&self) -> bool {
+        self.has_waker.swap(true, Ordering::AcqRel)
+    }
+}
+
+cfg_io_source! {
+    impl Selector {
+        #[cfg(debug_assertions)]
+        pub fn id(&self) -> usize {
+            self.state.id
+        }
+    }
+}
+
+/// Interface to poll.
+#[derive(Debug)]
+struct SelectorState {
```

Comment:
I might be remembering this wrong, but didn't the `polling` implementation also have an optional `timerfd` object to support sub-millisecond timeouts?

Reply:
I've copy & pasted most of this straight from `polling`, which doesn't have a `timerfd`. That being said, adding a `timerfd` is certainly possible. I however would like to not introduce it in this PR as to not further increase complexity.

For the time being you can technically add your own `timerfd` just fine to the registry and supply a `None` timeout.